### PR TITLE
[FLINK-13325][test] Add test to guard the fix for FLINK-13249

### PR DESCRIPTION
**Pull request by StefanRRichter**
## What is the purpose of the change

This PR is only adding the missing test case to guard our fix for FLINK-13249.

### CODE REVIEW

**Comment by StefanRRichter** [comment]
I created a test for the problem that is hopefully a good compromise between testing a rather low-level detail and not going to super deep in the IT case territory. Wdyt guys?

CC @zhijiangW

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskTest.java [hunk 1]
```diff
@@ -148,6 +120,7 @@ public void teardown() throws Exception {
 	public void testRegularExecution() throws Exception {
 		final QueuedNoOpTaskManagerActions taskManagerActions = new QueuedNoOpTaskManagerActions();
 		final Task task = createTaskBuilder()
+			.setInvokable(TestInvokableCorrect.class)
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskTest.java]: This change might be a separate pre hotfix commit.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskTest.java [hunk 2]
```diff
@@ -1013,197 +986,8 @@ public void notifyFatalError(String message, Throwable cause) {
 	//  helper functions
 	// ------------------------------------------------------------------------
 
-	private void setInputGate(Task task, SingleInputGate inputGate) {
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskTest.java]: nit: this legacy code remove could also be a separate commit.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TestTaskBuilder.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+	private Executor executor;
+	private Configuration taskManagerConfig;
+	private ExecutionConfig executionConfig;
+	private Collection<PermanentBlobKey> requiredJarFileBlobKeys;
+	private Collection<ResultPartitionDeploymentDescriptor> resultPartitions = Collections.emptyList();
+	private Collection<InputGateDeploymentDescriptor> inputGates = Collections.emptyList();
+
+	{
+		invokable = AbstractInvokable.class;
+		taskManagerActions = mock(TaskManagerActions.class);
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TestTaskBuilder.java]: It might use `NoOpTaskManagerActions` instead to avoid mock after refactoring.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TestTaskBuilder.java [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+		libraryCacheManager = mock(LibraryCacheManager.class);
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TestTaskBuilder.java]: ditto: `ContextClassLoaderLibraryCacheManager.INSTANCE` to avoid mock.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TestTaskBuilder.java [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+		when(libraryCacheManager.getClassLoader(any(JobID.class))).thenReturn(getClass().getClassLoader());
+
+		consumableNotifier = new NoOpResultPartitionConsumableNotifier();
+		partitionProducerStateChecker = mock(PartitionProducerStateChecker.class);
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TestTaskBuilder.java]: We could make `NoOpPartitionProducerStateChecker` public and use this.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TestTaskBuilder.java [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+		final TaskInformation taskInformation = new TaskInformation(
+			jobVertexId,
+			"Test Task",
+			1,
+			1,
+			invokable.getName(),
+			new Configuration());
+
+		final BlobCacheService blobCacheService = new BlobCacheService(
+			mock(PermanentBlobCache.class),
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TestTaskBuilder.java]: ditto: avoid mock if possible

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TestTaskBuilder.java [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+			jobInformation,
+			taskInformation,
+			executionAttemptId,
+			new AllocationID(),
+			0,
+			0,
+			resultPartitions,
+			inputGates,
+			0,
+			mock(MemoryManager.class),
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TestTaskBuilder.java]: ditto: `MemoryManager` and `IOManager`

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TestTaskBuilder.java [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+			new TestingTaskManagerRuntimeInfo(taskManagerConfig),
+			taskMetricGroup,
+			consumableNotifier,
+			partitionProducerStateChecker,
+			executor);
+	}
+
+	public static void setTaskState(Task task, ExecutionState state) {
+		try {
+			Field f = Task.class.getDeclaredField("executionState");
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TestTaskBuilder.java]: Another option is that we make the method `Task#transitionState` as package private.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannelTest.java [hunk 9]
```diff
@@ -1081,6 +1094,79 @@ private RemoteInputChannel createRemoteInputChannel(
...
+
+		// The entry point to that will lead us into blockingConnectionManager#createPartitionRequestClient(...).
+		inputGate.requestPartitions();
+
+		simulatedNetworkThread.join();
+
+		Assert.assertFalse(
+			"Test ended by timeout or interruption - this indicates that the network thread was blocked.",
+			timedOutOrInterrupted.get());
+	}
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannelTest.java]: Might be better to close `NettyShuffleEnvironment` and `SingleInputGate` finally.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannelTest.java [hunk 10]
```diff
@@ -1081,6 +1094,79 @@ private RemoteInputChannel createRemoteInputChannel(
 			.buildRemoteAndSetToGate(inputGate);
 	}
 
+	/**
+	 * Test to guard against FLINK-13249.
+	 */
+	@Test
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannelTest.java]: Might be better not to put this test among utility methods.

**Comment by StephanEwen** [comment]
I validated that the test indeed fails without the original fix.

Will address @zhijiangW's comments and merge this.
